src/kubernetes/reflector.rs

```diff
@@ -447,581 +446,574 @@ mod tests {
-            result.unwrap_err();
+        // The only way reflector completes is with an error, but that's ok.
+        // In tests we make it exit with an error to complete the test.
+        result.unwrap_err();
 
-            // Explicitly drop the reflector at the very end.
-            drop(reflector);
-        })
+        // Explicitly drop the reflector at the very end.
+        drop(reflector);
     }
 
     /// Test that the delayed delete works accordingly.
-    #[test]
-    fn test_delayed_deletes() {
-        test_util::trace_init();
-        test_util::block_on_std(async move {
-            // Freeze time.
-            tokio::time::pause();
-
-            // Prepare state.
-            let (state_events_tx, mut state_events_rx) = mpsc::channel(0);
-            let (mut state_actions_tx, state_actions_rx) = mpsc::channel(0);
-            let state_writer = state::mock::Writer::new(state_events_tx, state_actions_rx);
-            let state_writer = state::instrumenting::Writer::new(state_writer);
-            let deletion_delay = Duration::from_secs(600);
-            let state_writer = state::delayed_delete::Writer::new(state_writer, deletion_delay);
-
-            // Prepare watcher.
-            let (watcher_events_tx, mut watcher_events_rx) = mpsc::channel(0);
-            let (mut watcher_invocations_tx, watcher_invocations_rx) = mpsc::channel(0);
-            let watcher = MockWatcher::<Pod>::new(watcher_events_tx, watcher_invocations_rx);
-            let watcher = InstrumentingWatcher::new(watcher);
-
-            // Prepare reflector.
-            let mut reflector =
-                Reflector::new(watcher, state_writer, None, None, Duration::from_secs(1));
-
-            // Run test logic.
-            let logic = tokio::spawn(async move {
-                // Wait for watcher to request next invocation.
-                assert!(matches!(
-                    watcher_events_rx.next().await.unwrap(),
-                    mock_watcher::ScenarioEvent::Invocation(_)
-                ));
-
-                // Provide watcher with a new stream.
-                let (mut watch_stream_tx, watch_stream_rx) = mpsc::channel(0);
-                watcher_invocations_tx
-                    .send(mock_watcher::ScenarioActionInvocation::Ok(watch_stream_rx))
-                    .await
-                    .unwrap();
-
-                // Wait for watcher to request next item from the stream.
-                assert_eq!(
-                    watcher_events_rx.next().await.unwrap(),
-                    mock_watcher::ScenarioEvent::Stream
-                );
-
-                // Send pod addition to a stream.
-                watch_stream_tx
-                    .send(mock_watcher::ScenarioActionStream::Ok(WatchResponse::Ok(
-                        WatchEvent::Added(make_pod("uid0", "10")),
-                    )))
-                    .await
-                    .unwrap();
+    #[tokio::test(threaded_scheduler, core_threads = 2)]
```

Comment:
Why `core_threads = 2` here but not in other tests?

The amount of the threads (or even the runtime kind) should not be relevant for these tests - if they fail in any of the possible configurations - it's a bug.

Reply:
`threaded_scheduler` / `core_threads = 2` added because otherwise test hung up. I completely understand that we should able to use `tokio::test` in the default configuration and want to address this issue in future PRs.

Reply:
You mind creating an issue for this, and other things like this you've encountered? It's hard to tell if the change was problematic from the review. This one is easy to spot, but I assume there are other tiny details without which tests fail but and that we should look into more closely.

Reply:
I added an update plan in https://github.com/timberio/vector/issues/3478#issue-680213492. Sorry that did not mention all ths things before :disappointed:

Reply:
I'll create an issue for the k8s reflector tests in particular for myself. It looks like this change introduces the behaviour I didn't account for, and I'd like to ensure tests remain reliable in the long run.